Add validation and helper for rpc response - Closes#938

ishantiw opened the pull request:
### Description

Add util functions to validate individual incoming values and helper functions to validate peer objects. Add related tests.

Closes #938

jondubois commented on packages/lisk-p2p/test/unit/utils.ts:
```diff
@@ -0,0 +1,90 @@
+		const correctIp = '127.0.0.1';
+		const IpWithCharacters = '127k.0.0.1';
+		const IpWithLessBlocks = '127.0.0';
+		const IpWithSpaces = '127.0.0. 1';
```
We don't need so many simple cases (one or two is enough?).

A more useful case would be related to IPv4 vs IPv6 formats.

Another case would be; what if the ip provided is not a string but an object or null.

SargeKhan commented on packages/lisk-p2p/src/utils.ts:
```diff
@@ -0,0 +1,73 @@
+const MIN_VAL_PORT = 1;
+// We disabled no-any because we want to validate the values in response from the network that could be any
+/* tslint:disable: no-any */
+export const validateIp = (ipAddress: any): boolean => {
```
maybe it's better to use some schema checking library?

ishantiw replied:
@SargeKhan should we use https://www.npmjs.com/package/@types/validator for validator.js?

ishantiw replied:
@jondubois what do you think?

SargeKhan replied:
Are these the only validations which will be required in p2p library or do you see the validators list expanding?

ishantiw replied:
I think there will be a few more validations, but there will not be many. But there are some validations which are tricky, like ipv4/ipv6 which is better handled by the library. I just have validation for ipv4 right now and Jon said to have it for ipv6 too

shuse2 replied:
For schema check, we use https://github.com/epoberezkin/ajv.
For IP, I think core was using some library, which we can use it too.

ishantiw replied:
So, in core, `ip` is used in module and logic and `net` in app.js. which one to use or we should use `validator.js` that covers everything?

shuse2 commented on packages/lisk-p2p/src/helpers.ts:
```diff
@@ -0,0 +1,40 @@
+/*
```
Prefer not to use `helper`.  something like `response_handler.ts`?

ishantiw commented:
Actually, I am mostly using the approach similar to what you use. It's just that for the validators and utils I felt this approach will look more clear for all input types. But like you said, we need to have consistency here. I can change to your approach for these tests if you guys agree.

shuse2 commented on packages/lisk-p2p/src/utils.ts:
```diff
@@ -0,0 +1,73 @@
+	return isValid;
+};
+
+export const validatePort = (port: any): boolean => {
```
probably this should be `port: unknown`

ishantiw replied:
if I use `unknown`, `const port = +port` will throw error `Object is of type 'unknown'.`

ishantiw replied:
this the file has been removed now and validation is done with an external lib

shuse2 commented on packages/lisk-p2p/src/utils.ts:
```diff
@@ -0,0 +1,73 @@
+	);
+};
+
+export const onlyDigits = (input: any): boolean => {
```
same as above

shuse2 commented on packages/lisk-p2p/src/validators.ts:
```diff
@@ -0,0 +1,39 @@
+import { onlyDigits, validateIp, validatePort } from './utils';
+
+interface ResponsePeerObject {
+	readonly height: string | number;
```
why is this expecting string or number? shouldn't it be one way or another?

ishantiw replied:
From the response, we can expect either number or string, and both could be valid, but after we sanitize the response then we instantiate peer object with the port as number type. Or should we only use, one type?

ishantiw replied:
addressed

shuse2 commented on packages/lisk-p2p/src/validators.ts:
```diff
@@ -0,0 +1,39 @@
+interface ResponsePeerObject {
+	readonly height: string | number;
+	readonly ip: string;
+	readonly wsPort: string | number;
```
same as above

shuse2 commented on packages/lisk-p2p/src/helpers.ts:
```diff
@@ -0,0 +1,42 @@
+	readonly wsPort: string | number;
+}
+
+export const instantiatePeerFromResponse = (peer: unknown): boolean | Peer => {
```
Maybe better not to return two different type

ishantiw commented:
I have moved all the validation stuff of rpc response to `response_handler.ts` and I'm using `validator.js` lib for validation which has validation for` ip4/6` and `port` with a lot of other validations so makes sense to use just one library for all the validations in p2p. Also, `validator.js` is very popular with no dependencies. 

I also refactored test cases according to the general approach we discussed. 

@jondubois @SargeKhan @shuse2 Please review it again.

jondubois commented on packages/lisk-p2p/test/unit/response_handler.ts:
```diff
@@ -0,0 +1,116 @@
+				return expect(checkPeerAddress(peer)).to.be.true;
+			});
+
+			it('should return false', () => {
```
We should try to be more descriptive with the names of the test cases. But I think this is a minor thing.

shuse2 commented on packages/lisk-p2p/src/response_handler.ts:
```diff
@@ -0,0 +1,78 @@
+
+export const instantiatePeerFromResponse = (
+	peer: unknown,
+): PeerConfig | Error => {
```
it's not returning error anywhere?

ishantiw replied:
it is returning at [response_handler.ts#L52](https://github.com/LiskHQ/lisk-elements/blob/062056245576654306d35340c94e58722534d334/packages/lisk-p2p/src/response_handler.ts#L52)

ishantiw replied:
Discussed with @jondubois to throw an error instead of returning an error to have consistency across the p2p lib

shuse2 commented on packages/lisk-p2p/src/response_handler.ts:
```diff
@@ -0,0 +1,78 @@
+
+	const peerConfig: PeerConfig = {
+		ipAddress: rawPeer.ip,
+		wsPort: +rawPeer.wsPort,
```
we want to handle port as number internally, but return it  as string?
why not make it consistent?

ishantiw replied:
but here I'm converting it into integer and returning it

shuse2 commented on packages/lisk-p2p/src/response_handler.ts:
```diff
@@ -0,0 +1,78 @@
+	const os = rawPeer.os && isAlpha(rawPeer.os.toString()) ? rawPeer.os : '';
+	const version =
+		rawPeer.version && isAlpha(rawPeer.version.toString())
+			? rawPeer.version
```
maybe we should throw if version is empty. (maybe this relates to hand shake).
Please confirm with the core

ishantiw replied:
I will check this with Jon and core team

ishantiw replied:
as discussed, I will use `semver` that we use in core to validate version and throw the error if invalid

shuse2 commented on packages/lisk-p2p/src/response_handler.ts:
```diff
@@ -0,0 +1,78 @@
+			? +rawPeer.height
+			: 1;
+
+	const peerConfig: PeerConfig = {
```
what's the difference between `RawPeer` and `PeerConfig`?

ishantiw replied:
actually, `RawPeer ` will be replaced with `ProtocolPeerInfo` in `p2p_types` which will come from @jondubois  PR. The only difference is that we want a different type for a peer info in response from the network and most of the values will be string, that will be later converted to specific types when creating their corresponding `PeerConfig`

ishantiw replied:
another reason of not using `PeerConfig` is that we are using `ipAddress` instead of `ip` in PeerConfig

jondubois replied:
`PeerConfig` and `ProtocolPeerInfo` have almost the same properties (except for `ip` vs `ipAddress`) but the main reason why we declared them as different types is to decouple internal logic from the external protocol.

If we decide to go with a consistent type for everything (external and internal data), then our internal logic will be more tightly coupled to the current protocol.

jondubois commented on packages/lisk-p2p/test/unit/response_handler.ts:
```diff
@@ -0,0 +1,170 @@
+		});
+
+		describe('for invalid peer response object', () => {
+			it('should return false for invalid peer values', () => {
```
You should update the text since it's now throwing instead of returning.

ishantiw replied:
addressed

shuse2 commented on packages/lisk-p2p/src/response_handler.ts:
```diff
@@ -0,0 +1,81 @@
+	readonly wsPort: string;
+}
+
+export const checkForValidPeerAddress = (
```
check and valid is similar meaning.
maybe `validPeerAddress` or `validatePeerAddress`?
